File: stocks.pl
#!/usr/bin/perl
use strict;
use warnings;
use JSON qw( decode_json );
use LWP::UserAgent;
use Term::ANSIScreen qw( cls );
#use Chart::Graph::Gnuplot qw(gnuplot);

system("cls");

sub build_endpoint {
	$_[1] =~ s/[\%]/%25/g; #this must be done first

	$_[1] =~ s/[\ ]/%20/g;
	$_[1] =~ s/[\!]/%21/g;
	$_[1] =~ s/[\"]/%22/g;
	$_[1] =~ s/[\#]/%23/g;
	$_[1] =~ s/[\$]/%24/g;
	$_[1] =~ s/[\&]/%26/g;
	$_[1] =~ s/[\']/%27/g;
	$_[1] =~ s/[\(]/%28/g;
	$_[1] =~ s/[\)]/%29/g;
	$_[1] =~ s/[\*]/%2A/g;
	$_[1] =~ s/[\+]/%2B/g;
	$_[1] =~ s/[\,]/%2C/g;
	$_[1] =~ s/[\-]/%2D/g;
	$_[1] =~ s/[\.]/%2E/g;
	$_[1] =~ s/[\/]/%2F/g;
	$_[1] =~ s/[\:]/%3A/g;
	$_[1] =~ s/[\;]/%3B/g;
	$_[1] =~ s/[\<]/%3C/g;
	$_[1] =~ s/[\=]/%3D/g;
	$_[1] =~ s/[\>]/%3E/g;
	$_[1] =~ s/[\?]/%3F/g;
	$_[1] =~ s/[\@]/%40/g;

	return sprintf("%s?q=%s&format=%s&diagnostics=%s&env=%s&callback=%s", @_); 
}

sub make_request {
	# set custom HTTP request header fields
	my $req = HTTP::Request->new(GET => $_[0]);
	$req->header('content-type' => 'application/json');
	$req->header('x-auth-token' => 'kfksj48sdfj4jd9d');

	my $resp = $_[1]->request($req);
	if ($resp->is_success) {
	    my $response = decode_json($resp->decoded_content);
	    my $quote = $response->{'query'}{'results'}{'quote'};
	    my %return = ('Name' => $quote->{'Name'},
	    			  'Bid'  => $quote->{'Bid'}, 
	    			  'Ask'  => $quote->{'Ask'}, 
	    			  'Time' => $response->{'query'}{'created'});
	    return %return;
	} 
	else {
	    print "HTTP GET error code: ", $resp->code, "\n";
	    print "HTTP GET error message: ", $resp->message, "\n";
	    exit 1; 
	}
}

my $ua       = LWP::UserAgent->new;
my $query    = "SELECT Name, Bid, Ask, LastTradePriceOnly FROM yahoo.finance.quotes WHERE symbol=\"TSLA\"";
my $env      = "store%3A%2F%2Fdatatables.org%2Falltableswithkeys";
my $endpoint = build_endpoint( "https://query.yahooapis.com/v1/public/yql", 
							   $query,   #YQL query
							   "json",   #format of response
							   "false",  #boolean to include diagnostic info
							   $env,     #
							   "" );     #callback


=begin comment
my @x_column;
my @y_column;

for ( $i=0; $i<1; $i++ ) {
	%response = make_request($endpoint);
	$x_column[$i] = $response{'Time'};
	$y_column[$i] = $response{'Bid'};

	gnuplot({'title' => 'foo'},
          [@x_column, @y_column ]);
}
=cut

$| = 1;
my %response = make_request($endpoint, $ua);
print $response{'Name'}."\n";
print "Bid : ".$response{'Bid'}."\n";
print "Ask : ".$response{'Ask'};

for ( ; ; ) {
	sleep 10;
	system("cls"); #cls();

	%response = make_request($endpoint, $ua);
	print $response{'Name'}."\n";
	print "Bid : ".$response{'Bid'}."\n";
	print "Ask : ".$response{'Ask'};
}

exit 0;

File: stocks.py
import urllib, urllib2
import json, re, math
import time, os
os.system('cls')

MAX_LENGTH   = 80
H_BUFFER     = 2
V_BUFFER     = 4
REFRESH_RATE = 2
WATCHLIST    = ['TSLA', 'GOOGL', 'GOOG', 'AAPL', 'OMCL', 'AAL', 'GPRO']


class Stock:
	data     = {}
	url      = 'https://query.yahooapis.com/v1/public/yql'
	sql_stmt = 'SELECT Name, Bid, Ask, LastTradePriceOnly FROM yahoo.finance.quotes WHERE symbol="{}"'

	def __init__(self, ticker):
		self.ticker              = ticker
		self.data['q']           = re.sub(r'\{\}', ticker, self.sql_stmt)
		self.data['format']      = 'json'
		self.data['diagnostics'] = 'false'
		self.data['env']         = 'store://datatables.org/alltableswithkeys'
		self.data['callback']    = ''
		self.full_url = self.url + '?' + urllib.urlencode(self.data)
		self.refresh_data()

	def refresh_data(self):
		response  = json.load(urllib2.urlopen(self.full_url))
		# TODO: add error checking
		self.name = response['query']['results']['quote']['Name']
		self.bid  = response['query']['results']['quote']['Bid']
		self.ask  = response['query']['results']['quote']['Ask']

	def get_name_string(self):
		return self.name + ' (' + self.ticker + ')'

	def get_bid_string(self):
		return "Bid : " + self.bid

	def get_ask_string(self):
		return "Ask : " + self.ask


stocks = []
for ticker in WATCHLIST:
	stocks.append(Stock(ticker))

while (1):
	output  = []
	longest = 0

	for stock in stocks:
		stock.refresh_data()
		if (longest < len(stock.get_name_string())): 
			longest = len(stock.get_name_string())
		if (longest < len(stock.get_bid_string())): 
			longest = len(stock.get_bid_string())
		if (longest < len(stock.get_ask_string())): 
			longest = len(stock.get_ask_string())

	columns = int(math.floor(MAX_LENGTH/float(longest+H_BUFFER)))
	#if ((columns*longest+(columns-1)*H_BUFFER) < MAX_LENGTH): 
	#	columns = columns + 1
	rows    = int(math.ceil(len(stocks)/float(columns)))

	for r in range(0, rows):
		for x in range(0, 3):
			output.append("")
		for c in range(0, columns):
			if (c+(r*columns) >= len(stocks)):
				break
			stock = stocks[c+(r*columns)]
			output[(0)+(r*3)] += stock.get_name_string()
			output[(1)+(r*3)] += stock.get_bid_string()
			output[(2)+(r*3)] += stock.get_ask_string()
			if not (c==(columns-1)):
				for x in range(0, (longest-len(stock.get_name_string()))+H_BUFFER):
					output[(0)+(r*3)] += " "
				for x in range(0, (longest-len(stock.get_bid_string()))+H_BUFFER):
					output[(1)+(r*3)] += " "
				for x in range(0, (longest-len(stock.get_ask_string()))+H_BUFFER):
					output[(2)+(r*3)] += " "

	os.system('cls')
	for l in range(0, len(output)):
		if (l % 3 == 0) and not (l == 0):
			for x in range(0, V_BUFFER):
				print ""
		print output[l]
	time.sleep(REFRESH_RATE)